### PR TITLE
Mark fields in Condition required

### PR DESCRIPTION
**Pull request by maplain**
Based on the recommendation in https://github.com/kubernetes/enhancements/pull/1624, make the following changes on GatewayConditon, GatewayClassCondition and ListenerCondition:
1. add +required for all fields;
2. make existing pointer fields non-pointers;

### CODE REVIEW

**Comment by jpeach** [comment]
Notes from office hours conversion:
- check for prior art across k/k for this change
- if we do this, do it for all the condition types

**Comment by robscott** [comment]
@maplain Thanks for this PR! And thanks to Tim for the clarification on how we should be using these types. Based on that, it looks like we should actually be headed in the other direction, removing `+optional` tags for these fields and ensuring they are not pointers. This should also be updated here: https://github.com/kubernetes-sigs/service-apis/blob/master/api/v1alpha1/gatewayclass_types.go#L149-L171. I can file a separate issue if these changes seem out of scope here.

**Comment by maplain** [comment]
Thanks @jpeach @thockin and @robscott!

I updated my PR to reflect changes on all Conditions: GatewayCondition, GatewayClassCondition and ListenerCondition. Changes include:
1. add `+required` for all fields;
2. make existing pointer fields non-pointers;

File: api/v1alpha1/gateway_types.go [hunk 1]
```diff
@@ -335,18 +335,22 @@ const (
...
+	// +required
 	Type GatewayConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=GatewayConditionType"`
 	// Status describes the current state of this condition. Can be "True",
 	// "False", or "Unknown".
+	// +required
 	Status core.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
 	// Message is a human-understandable message describing the condition.
-	// +optional
+	// This field may be empty.
+	// +required
```
Comment (robscott) [inline api/v1alpha1/gateway_types.go:346]: I know this is how https://github.com/kubernetes/enhancements/pull/1624 recommends defining this field, but it sure feels less than ideal. I've left a [comment](https://github.com/kubernetes/enhancements/pull/1624) on that PR to try to get some clarification on the rationale.

Comment (maplain) [reply api/v1alpha1/gateway_types.go:346]: sounds good. my understanding is there is no use case to differentiate an empty string from no input from the user in reality.

I'll follow up on that thread too.

**Comment by bowei** [comment]
We may have to revisit this given the discussion on the linked PR. I'm going to merge this for now as it is consistent with the current proposal.

/lgtm
/approve

**Pull request merged**